delta_go/liberty_tracker.py

```diff
@@ -0,0 +1,164 @@
+import copy
+from typing import Dict, Optional, Set
+
+import numpy as np
+
+from go_base import Group, find_reached, place_stones
+from utils import BLACK, BOARD_SIZE, EMPTY, FILL, MISSING_GROUP_ID, NEIGHBORS, WHITE
+
+
+class LibertyTracker:
+    @staticmethod
+    def from_board(board: np.ndarray) -> "LibertyTracker":
+        board = np.copy(board)
+        curr_group_id = 0
+        lib_tracker = LibertyTracker()
+        for color in (WHITE, BLACK):
+            while color in board:
+                curr_group_id += 1
+                found_color = np.where(board == color)
+                coord = found_color[0][0], found_color[1][0]
+                chain, reached = find_reached(board, coord)
+                liberties = frozenset(r for r in reached if board[r] == EMPTY)
+                new_group = Group(curr_group_id, frozenset(chain), liberties, color)
+                lib_tracker.groups[curr_group_id] = new_group
+                for s in chain:
+                    lib_tracker.group_index[s] = curr_group_id
+                place_stones(board, FILL, chain)
+
+        lib_tracker.max_group_id = curr_group_id
+
+        liberty_counts = np.zeros([BOARD_SIZE, BOARD_SIZE], dtype=np.uint8)
+        for group in lib_tracker.groups.values():
+            num_libs = len(group.liberties)
+            for s in group.stones:
+                liberty_counts[s] = num_libs
+        lib_tracker.liberty_cache = liberty_counts
+
+        return lib_tracker
+
+    def __init__(
+        self,
+        group_index: Optional[np.ndarray] = None,
+        groups: Optional[Dict] = None,
+        liberty_cache: Optional[np.ndarray] = None,
+        max_group_id: int = 1,
+    ):
+        # group_index: a NxN numpy array of group_ids. -1 means no group
+        # groups: a dict of group_id to groups
+        # liberty_cache: a NxN numpy array of liberty counts
+        self.group_index = (
+            group_index
+            if group_index is not None
+            else -np.ones([BOARD_SIZE, BOARD_SIZE], dtype=np.int32)
+        )
+        self.groups = groups or {}
+        self.liberty_cache = (
+            liberty_cache
+            if liberty_cache is not None
+            else np.zeros([BOARD_SIZE, BOARD_SIZE], dtype=np.uint8)
+        )
+        self.max_group_id = max_group_id
+
+    def __deepcopy__(self, memo):
+        new_group_index = np.copy(self.group_index)
+        new_lib_cache = np.copy(self.liberty_cache)
+        # shallow copy
+        new_groups = copy.copy(self.groups)
+        return LibertyTracker(
+            new_group_index,
+            new_groups,
+            liberty_cache=new_lib_cache,
+            max_group_id=self.max_group_id,
+        )
+
+    def add_stone(self, color, c):
+        assert self.group_index[c] == MISSING_GROUP_ID
+        captured_stones = set()
+        opponent_neighboring_group_ids = set()
+        friendly_neighboring_group_ids = set()
+        empty_neighbors = set()
```

Comment:
Types on this whole class would be 10/10. Also an explanation of what the class is in a docstring would be useful.

As far as I can remember, the `liberty_cache` property is a np array of the number of liberties each square has (including liberties of the group that square is a part of )

Reply:
Have added a docstring and a ticket for the typing